File: tasks/builtin/bug-fix-03/starter.py
import threading
import time


class ThreadSafeCounter:
    def __init__(self):
        self._value = 0
        self._lock = threading.Lock()

    def increment(self):
        current = self._value
        time.sleep(0.00001)
        self._value = current + 1

    def decrement(self):
        with self._lock:
            self._value -= 1

    def get(self):
        return self._value

    def reset(self):
        with self._lock:
            self._value = 0

    def increment_by(self, n: int):
        current = self.get()
        with self._lock:
            self._value = current + n

File: tasks/builtin/bug-fix-03-go/starter.go
package solution

import (
	"sync"
	"time"
)

type ThreadSafeCounter struct {
	value int
	mu    sync.Mutex
}

func NewThreadSafeCounter() *ThreadSafeCounter {
	return &ThreadSafeCounter{value: 0}
}

func (c *ThreadSafeCounter) Increment() {
	current := c.value
	time.Sleep(10 * time.Microsecond)
	c.value = current + 1
}

func (c *ThreadSafeCounter) Decrement() {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.value--
}

func (c *ThreadSafeCounter) Get() int {
	return c.value
}

func (c *ThreadSafeCounter) Reset() {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.value = 0
}

func (c *ThreadSafeCounter) IncrementBy(n int) {
	current := c.Get()
	c.mu.Lock()
	defer c.mu.Unlock()
	c.value = current + n
}
